### ISSUE TITLE
Killing an isolate with an active timer may stop *all* timers

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#28115]
If an isolate creates a periodic timer, and that isolate is then killed, *all* timers in the VM will stop. For example:

```dart
import 'dart:async';
import 'dart:isolate';

main() async {
  var port = new ReceivePort();
  var isolate = await Isolate.spawnUri(new Uri.dataFromString("""
    import 'dart:async';
    import 'dart:isolate';

    main(_, SendPort sendPort) {
      new Timer.periodic(new Duration(milliseconds: 100), (_) {
        sendPort.send(null);
        print("inner tick");
      });
    }
  """), [], port.sendPort);

  new Timer(new Duration(seconds: 1), () => print("outer tick"));
  await port.first;
  isolate.kill();
}
```

**Comment by nex3** [comment]
Marking as P1 because this means that any code that kills isolates (including `test` and pub) can potentially deadlock the VM.

**Comment by floitschG** [comment]
There is no need for periodic timers.

``` dart
import 'dart:async';
import 'dart:isolate';

main() async {
  var port = new ReceivePort();
  var isolate = await Isolate.spawnUri(new Uri.dataFromString("""
    import 'dart:async';
    import 'dart:isolate';

    main(_, SendPort sendPort) {
      sendPort.send(null);
      new Timer(new Duration(milliseconds: 100), () {
        sendPort.send(null);
        print("inner tick");
      });
    }
  """), [], port.sendPort);

  new Timer(new Duration(seconds: 1), () => print("outer tick"));
  await port.first;
  print("received tick");
  isolate.kill();
}
```

I tried to debug, but at some point it enters the VM.
I found a patch that solves the issue, but I'm not sure it's correct.

``` diff
diff --git a/runtime/bin/eventhandler_linux.cc b/runtime/bin/eventhandler_linux.cc
index 8b35818..52c8ab9 100644
--- a/runtime/bin/eventhandler_linux.cc
+++ b/runtime/bin/eventhandler_linux.cc
@@ -357,6 +357,16 @@ void EventHandlerImplementation::HandleEvents(struct epoll_event* events,
       if (timeout_queue_.HasTimeout()) {
         DartUtils::PostNull(timeout_queue_.CurrentPort());
         timeout_queue_.RemoveCurrent();
+
+        struct itimerspec it;
+        memset(&it, 0, sizeof(it));
+        if (timeout_queue_.HasTimeout()) {
+          int64_t millis = timeout_queue_.CurrentTimeout();
+          it.it_value.tv_sec = millis / 1000;
+          it.it_value.tv_nsec = (millis % 1000) * 1000000;
+        }
+        VOID_NO_RETRY_EXPECTED(
+            timerfd_settime(timer_fd_, TFD_TIMER_ABSTIME, &it, NULL));
       }
     } else {
       DescriptorInfo* di =
```

(This functionality should probably be shared with the other call to `timerfd_settime`).

**Comment by floitschG** [comment]
@a-siva : small ping. I hope we could get this into 1.22.

**Comment by rmacnak-google** [comment]
60a7160273f881304a03875cb9c16e8d7b791ab9

**Comment by nex3** [comment]
This is a major bug fix--can we back-port it to 1.24?
